samples/AspNetCoreSseServer/AspNetCoreSseMcpTransport.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System.Buffers;
+using System.Net.ServerSentEvents;
+using System.Text.Json;
+using System.Threading.Channels;
+using McpDotNet.Protocol.Messages;
+using McpDotNet.Protocol.Transport;
+using McpDotNet.Utils.Json;
+
+namespace AspNetCoreSseServer;
+
+public class AspNetCoreSseMcpTransport(Stream sseResponseStream, ILogger<AspNetCoreSseMcpTransport> logger) : ITransport
+{
+    [ThreadStatic]
+    private static Utf8JsonWriter? _jsonWriter;
+
+    private readonly Channel<IJsonRpcMessage> _incomingChannel = CreateSingleItemChannel<IJsonRpcMessage>();
+    private readonly Channel<SseItem<IJsonRpcMessage?>> _outgoingSseChannel = CreateSingleItemChannel<SseItem<IJsonRpcMessage?>>();
+
+    private Task? _sseWriteTask;
+
+    public bool IsConnected => _sseWriteTask?.IsCompleted == false;
+
+    public Task RunAsync(CancellationToken cancellationToken)
+    {
+        static void WriteJsonRpcMessageToBuffer(SseItem<IJsonRpcMessage?> item, IBufferWriter<byte> writer)
+        {
+            if (item.EventType == "endpoint")
+            {
+                writer.Write("/message"u8);
+                return;
+            }
+
+            JsonSerializer.Serialize(GetUtf8JsonWriter(writer), item.Data, JsonSerializerOptionsExtensions.DefaultOptions);
+        }
+
+        // The very first SSE event isn't really an IJsonRpcMessage, but there's no API to write a single item of a different type,
+        // so we fib and special-case the "endpoint" event type in the formatter.
+        _outgoingSseChannel.Writer.TryWrite(new SseItem<IJsonRpcMessage?>(null, "endpoint"));
+
+        var sseItems = _outgoingSseChannel.Reader.ReadAllAsync(cancellationToken);
+        return _sseWriteTask = SseFormatter.WriteAsync(sseItems, sseResponseStream, WriteJsonRpcMessageToBuffer, cancellationToken);
```

Comment:
@eiriktsarpalis It was a bit awkward writing the first "endpoint" event using the new `SseFormatter` API. Unlike the "message" events, it is not JSON. I could have used the `IAsyncEnumerable<SseItem<string>>` overload of course, but I wanted to avoid the intermediate string allocations.

This is why I special cased `item.EventType == "endpoint"` in `WriteJsonRpcMessageToBuffer` to not write JSON.  I'm guessing I could have first written a single-element `IAsyncEnumerable<SseItem<string>>` with just the endpoint event before writing the `IAsyncEnumerable<SseItem< IJsonRpcMessage >>` instead, but I think it would have been nicer if I could write single events without initializing a collection.

Reply:
Yeah, it becomes awkward when multiple encodings are mixed in the stream. This approach seems good enough to me.